README.md
p5-regexp-optimizer
===================

Regexp::Optimizer - optimizes regular expressions

SYNOPSIS
--------

    use Regexp::Optimizer;
    my $o  = Regexp::Optimizer->new->optimize(qr/foobar|fooxar|foozap/);
    # $re is now qr/foo(?:[bx]ar|zap)/


Changes
#
# $Id: Changes,v 0.22 2013/02/26 04:51:15 dankogai Exp dankogai $
#
$Revision: 0.22 $ $Date: 2013/02/26 04:51:15 $
! t/02-optimizer.t
  Since qr// happens compile time, SKIP: is not good enough to make
  older perls (which do not support named captures) to make them skip.

$Revision: 0.22 $ $Date: 2013/02/26 04:51:15 $
! lib/Regexp/Optimizer.pm t/02-optimizer.t
  + Clearner Codes
  + More Tests
  + More Documents

0.20 2013/02/23 13:43:59
! *
  Completely rewritten.
  + uses Regexp::Assemble to optimize alteration
  + uses ??{CODE} to parse nested parens
  Thanks to these changes, this module is now much simpler.

0.16 2013/02/20 17:54:59
! lib/Regexp/Optimizer.pm lib/Regexp/List.pm
  Marked obsolete.  Use Regexp::Assemble instead.

0.15 2004/12/05 16:07:34
! lib/Regexp/Optimizer.pm lib/Regexp/List.pm
  Pod fixed accordingly to RT # 8733
  http://rt.cpan.org/NoAuth/Bug.html?id=8733

0.14 2004/11/05 12:44:48
! t/02-list.t
  Addressed test failure that was raised by Perl 5.8.5
  http://rt.cpan.org/NoAuth/Bug.html?id=8165

0.13 2004/05/08 05:55:35
! lib/Regexp/Optimizer.pm
  Document bug, reported by Frederick in the mail below, fixed.
! Makefile.PL
! Regexp/ -> lib/Regexp/
  Module hierarchy realigned so cygwin is happy.  
  Reported by Frederick Weiland
  <30F579C44E1598429167C5E20E307BA19D807D@atlmsg01.raremedium.net>

0.12 2004/05/04 17:12:14
! Regexp/Optimizer.pm
  Perl 5.8.4 and later corrected a bug so it is fatal to go
    my $x = qr{ ... (??{ $x }) ... };
  under "use strict".  Unfortunately Regexp/Optimizer.pm had two
  occurances thereof.  Now fixed.

0.11 2004/05/03 15:09:14
! Regexp/List.pm
  Deep recursion addressed and corrected by yoz@yoz.com
  https://rt.cpan.org/Ticket/Display.html?id=4937

0.10 2003/06/02 20:11:07
! *
  Version jumps to 0.10 do to radical changes, especially in
  $Regexp::Optimize.
! Regexp/Optimizer.pm t/02-list.t t/03-utf8.t
  Aagh!  Why didn't I come up with such a simple idea !?
  ->optimize is completely rewritten.
  t/*.t is streamlined to reflect changes.
  Now ->optimize() *really* optimizes even for
  qr/(?:1|12)|123/.  Used to return qr/1(?:2?|23)/ but now it
  returns qr/1(?:23?)?/o
! Regexp/Optimizer.pm  Regexp/List.pm
  POD enhanced.

0.02 2003/06/01 00:11:26
! Regexp/Optimizer.pm  Regexp/List.pm
  * Lots of bug fixes regarding $o->optimize() for nested parens
! t/02-list.t
  Test data for #21 corrected
! t/03-utf8.t
  buggy SKIP: sections fixed.

0.01 2003/05/31 10:44:41
+ *
  0th release


lib/Regexp/List.pm
package Regexp::List;

use 5.008001;
use strict;
use warnings FATAL => 'all';
use Regexp::Assemble;
our $VERSION = sprintf "%d.%02d", q$Revision: 0.20 $ =~ /(\d+)/g;

sub new {
    bless \my $dummy, shift;
};

sub list2re {
    my $self = shift;
    Regexp::Assemble->new->add(@_)->re;
};

1; # End of Regexp::List

__END__

=head1 NAME

Regexp::List - Assemble multiple Regular Expressions into a single RE

=head1 VERSION

$Id: List.pm,v 0.20 2013/02/23 13:43:59 dankogai Exp dankogai $

=head1 DEPRECATED

use L<Regexp::Assemble> instead.

=head1 SYNOPSIS

  use Regexp::List;
  my $rl = Regexp::List->new();
  my @list = ( 'ab+c', 'ab+-',  'a\w\d+', 'a\d+' );
  print $rl->list2re(@list);
  # Regexp::Asssemble->new->add(@list);

=head1 DESCRIPTION

This module exists just for the sake of compatibility w/ version 0.16
and below.

=over 2

=item new

Just a stub.

=item list2re

Simply does:

  Regexp::Asssemble->new->add(@list);

=back

=head1 SEE ALSO

L<Regexp::Optimizer>, L<Regexp::Assemble>

=head1 AUTHOR

Dan Kogai, C<< <dankogai+cpan at gmail.com> >>

=head1 BUGS

Please report any bugs or feature requests to C<bug-regexp-optimizer at rt.cpan.org>, or through
the web interface at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Regexp-Optimizer>.  I will be notified, and then you'll
automatically be notified of progress on your bug as I make changes.

=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc Regexp::Optimizer


You can also look for information at:

=over 4

=item * RT: CPAN's request tracker (report bugs here)

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=Regexp-Optimizer>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/Regexp-Optimizer>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/Regexp-Optimizer>

=item * Search CPAN

L<http://search.cpan.org/dist/Regexp-Optimizer/>

=back

=head1 LICENSE AND COPYRIGHT

Copyright 2013 Dan Kogai.

This program is free software; you can redistribute it and/or modify it
under the terms of the the Artistic License (2.0). You may obtain a
copy of the full license at:

L<http://www.perlfoundation.org/artistic_license_2_0>

Any use, modification, and distribution of the Standard or Modified
Versions is governed by this Artistic License. By using, modifying or
distributing the Package, you accept this license. Do not use, modify,
or distribute the Package, if you do not accept this license.

If your Modified Version has been derived from a Modified Version made
by someone other than you, you are nevertheless required to ensure that
your Modified Version complies with the requirements of this license.

This license does not grant you the right to use any trademark, service
mark, tradename, or logo of the Copyright Holder.

This license includes the non-exclusive, worldwide, free-of-charge
patent license to make, have made, use, offer to sell, sell, import and
otherwise transfer the Package with respect to any patent claims
licensable by the Copyright Holder that are necessarily infringed by the
Package. If you institute patent litigation (including a cross-claim or
counterclaim) against any party alleging that the Package constitutes
direct or contributory patent infringement, then this Artistic License
to you shall terminate on the date that such litigation is filed.

Disclaimer of Warranty: THE PACKAGE IS PROVIDED BY THE COPYRIGHT HOLDER
AND CONTRIBUTORS "AS IS' AND WITHOUT ANY EXPRESS OR IMPLIED WARRANTIES.
THE IMPLIED WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR
PURPOSE, OR NON-INFRINGEMENT ARE DISCLAIMED TO THE EXTENT PERMITTED BY
YOUR LOCAL LAW. UNLESS REQUIRED BY LAW, NO COPYRIGHT HOLDER OR
CONTRIBUTOR WILL BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, OR
CONSEQUENTIAL DAMAGES ARISING IN ANY WAY OUT OF THE USE OF THE PACKAGE,
EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

=cut


lib/Regexp/Optimizer.pm
package Regexp::Optimizer;

use 5.008001;
use strict;
use warnings FATAL => 'all';
use Regexp::Assemble;
our $VERSION = sprintf "%d.%02d", q$Revision: 0.22 $ =~ /(\d+)/g;

my $re_nested;
$re_nested = qr{
  \(                   # open paren
  ((?:                 # start capture  
    (?>[^()]+)       | # Non-parens w/o backtracking or ...
    (??{ $re_nested }) # Group with matching parens
  )*)                  # end capture
  \)                   # close paren
}msx;

my $re_optimize = qr{(?<=[^\\])\|}ms;

sub new {
    my $class = shift;
    bless {@_}, $class;
}

sub _assemble {
    my $str = shift;
    return $str if $str !~ $re_optimize;
    if ( $str !~ m/[(]/ms ) {
        my $ra = Regexp::Assemble->new();
        $ra->add( split m{[|]}, $str );
        return $ra->as_string;
    }
    $str =~ s{$re_nested}{
        no warnings 'uninitialized';
        my $sub = $1;
        if ($sub =~ m/\A\?(?:[\?\{\(PR]|[\+\-]?[0-9])/ms) {
            "($sub)";  # (?{CODE}) and like ruled out
        }else{
            my $mod = ($sub =~ s/\A\?//) ? '?' : '';
            if ($mod) {
                $sub =~ s{\A(
                              [\w\^\-]*: | # modifier
                              [<]?[=!]   | # assertions
                              [<]\w+[>]  | # named capture
                              [']\w+[']  | # ditto
                              [|]          # branch reset
                          )
                     }{}msx;
                $mod .= $1;
            }
            '(' . $mod . _assemble($sub) . ')'
        }
    }msxge;
    $str;
}

sub as_string {
    my ( $self, $str ) = @_;
    return $str if $str !~ $re_optimize;
    my ($mod) = ($str =~ m/\A\(\?(.*?):/);
    if ( $mod =~ /x/ ) {
        $str =~ s{^\s+}{}mg;
        $str =~ s{(?<=[^\\])\s*?#.*?$}{}mg;
        $str =~ s{\s+[|]\s+}{|}mg;
        $str =~ s{(?:\r\n?|\n)}{}msg;
        $str =~ s{[ ]+}{ }msgx;
        # warn $str;
    }
    # escape all occurance of '\(' and '\)'
    $str =~ s/\\([\(\)])/sprintf "\\x%02x" , ord $1/ge;
    _assemble($str);
}

sub optimize {
    my $self = shift;
    my $re   = $self->as_string(shift);
    qr{$re};
}

1; # End of Regexp::Optimizer

__END__

=head1 NAME

Regexp::Optimizer - optimizes regular expressions

=head1 VERSION

$Id: Optimizer.pm,v 0.22 2013/02/26 04:51:15 dankogai Exp dankogai $

=head1 SYNOPSIS

  use Regexp::Optimizer;
  my $o  = Regexp::Optimizer->new->optimize(qr/foobar|fooxar|foozap/);
  # $re is now qr/foo(?:[bx]ar|zap)/

=head1 EXPORT

none.

=head1 SUBROUTINES/METHODS

=head2 new

Makes a new optimizer instance.

  my $ro = Regexp::Optimizer->new;

=head2 optimize

Does the optimization.

  my $re = qr/foobar|fooxar|foozap/;
  $re = $ro->optimize($re);

If it is already optimized -- no alteration in the regexp, it is
practically an identity function which simply returns an argument.

If not, it dissasembles the regexp, feeds it to L<Regexp::Assemble>,
and reassembles the result.

=head2 as_string

Same as C<optimize()> but returns a string instead of regexp object.

=head1 CAVEAT

=head2 ??{CODE} used

This module depends on the C<??{CODE}> regexp construct which is still
considered experimental as of Perl 5.16.

=head2 not idempotent

If you feed the regexp that is already optimized, the resulting regexp
may not necessarily the same -- usually you get duplicate C<(?:)>:

    my $re = qr/foobar|fooxar|foozap/;
    $re = $ro->optimize($re); # qr/foo(?:[bx]ar|zap)/
    $re = $ro->optimize($re); # qr/foo(?:(?:[bx]ar|zap))/

=head1 SEE ALSO

L<Regexp::Assemble>, L<perlre>

=head1 AUTHOR

Dan Kogai, C<< <dankogai+cpan at gmail.com> >>

=head1 BUGS

Please report any bugs or feature requests to C<bug-regexp-optimizer at rt.cpan.org>, or through
the web interface at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Regexp-Optimizer>.  I will be notified, and then you'll
automatically be notified of progress on your bug as I make changes.

=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc Regexp::Optimizer

You can also look for information at:

=over 4

=item * RT: CPAN's request tracker (report bugs here)

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=Regexp-Optimizer>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/Regexp-Optimizer>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/Regexp-Optimizer>

=item * Search CPAN

L<http://search.cpan.org/dist/Regexp-Optimizer/>

=back

=head1 ACKNOWLEDGEMENTS

=head1 LICENSE AND COPYRIGHT

Copyright 2013 Dan Kogai.

This program is free software; you can redistribute it and/or modify it
under the terms of the the Artistic License (2.0). You may obtain a
copy of the full license at:

L<http://www.perlfoundation.org/artistic_license_2_0>

Any use, modification, and distribution of the Standard or Modified
Versions is governed by this Artistic License. By using, modifying or
distributing the Package, you accept this license. Do not use, modify,
or distribute the Package, if you do not accept this license.

If your Modified Version has been derived from a Modified Version made
by someone other than you, you are nevertheless required to ensure that
your Modified Version complies with the requirements of this license.

This license does not grant you the right to use any trademark, service
mark, tradename, or logo of the Copyright Holder.

This license includes the non-exclusive, worldwide, free-of-charge
patent license to make, have made, use, offer to sell, sell, import and
otherwise transfer the Package with respect to any patent claims
licensable by the Copyright Holder that are necessarily infringed by the
Package. If you institute patent litigation (including a cross-claim or
counterclaim) against any party alleging that the Package constitutes
direct or contributory patent infringement, then this Artistic License
to you shall terminate on the date that such litigation is filed.

Disclaimer of Warranty: THE PACKAGE IS PROVIDED BY THE COPYRIGHT HOLDER
AND CONTRIBUTORS "AS IS' AND WITHOUT ANY EXPRESS OR IMPLIED WARRANTIES.
THE IMPLIED WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR
PURPOSE, OR NON-INFRINGEMENT ARE DISCLAIMED TO THE EXTENT PERMITTED BY
YOUR LOCAL LAW. UNLESS REQUIRED BY LAW, NO COPYRIGHT HOLDER OR
CONTRIBUTOR WILL BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, OR
CONSEQUENTIAL DAMAGES ARISING IN ANY WAY OUT OF THE USE OF THE PACKAGE,
EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.


=cut
